include/hc.hpp

```diff
@@ -117,6 +451,188 @@ inline uint64_t get_tick_frequency() {
+                                                                                  const Kernel&);
+
+    #if __KALMAR_ACCELERATOR__ == 2 || __KALMAR_CPU__ == 2
+        template<typename Kernel, int N>
+        friend
+        completion_future launch_cpu_task_async(const std::shared_ptr<Kalmar::KalmarQueue>&,
+                                                const Kernel&,
+                                                const extent<N>&);
+    #endif
+
+    // non-tiled parallel_for_each
+    // generic version
+    template<int N, typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<N>&,
+                                        const Kernel&);
+
+    // 1D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<1>&,
+                                        const Kernel&);
+
+    // 2D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<2>&,
+                                        const Kernel&);
+
+    // 3D specialization
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const extent<3>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 3D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<3>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 2D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<2>&,
+                                        const Kernel&);
+
+    // tiled parallel_for_each, 1D version
+    template <typename Kernel>
+    friend
+    completion_future parallel_for_each(const accelerator_view&,
+                                        const tiled_extent<1>&,
+                                        const Kernel&);
+
+    static constexpr unsigned int max_async_ops = 5u; // TODO: retrieve directly from HSA runtime.
+
+    std::shared_ptr<Kalmar::KalmarQueue> queue_;
+
+    explicit
+    accelerator_view(std::shared_ptr<Kalmar::KalmarQueue> queue)
+        : queue_{std::move(queue)} {}
+
+    template<typename InputIterator>
+    completion_future create_blocking_marker_(InputIterator f,
+                                              InputIterator l,
+                                              memory_scope scope,
+                                              std::true_type) const
+    {   // Precondition: readable_bounded_range(f, l) && non_empty(f, l).
```

Comment:
Right, it does not handle empty-ranges because it is an implementation detail, and the outer caller (create_blocking_marker) already deals with that itself. Properly asserting around this is a pretty great idea indeed; they will show up before final PR, thank you.